src/main/java/net/minecraftforge/network/HandshakeMessages.java

```diff
@@ -83,7 +90,18 @@ public static S2CModList decode(FriendlyByteBuf input)
             for (int x = 0; x < len; x++)
                 registries.add(input.readResourceLocation());
 
-            return new S2CModList(mods, channels, registries);
+            // Datapack Registries may or may not be sent in 1.18.2 due to netcode changes.
+            // TODO 1.19: Remove optionalness of datapack registry list in the mod list packet.
+            List<ResourceKey<? extends Registry<?>>> dataPackRegistries = new ArrayList<>();
+            if (input.isReadable())
+            {
+                len = input.readVarInt();
+                for (int x = 0; x < len; x++)
+                {
+                    dataPackRegistries.add(ResourceKey.createRegistryKey(input.readResourceLocation()));
+                }
```

Comment:
Could you instead use `FriendlyByteBuf#readCollection` here (which is the counterpart to `FriendlyByteBuf#writeCollection`)?

Reply:
That's a neat function! Will change in next push